### PR TITLE
Fix factoring bug (O(n) arange related)

### PR DESCRIPTION
**Pull request by patosai**
See #3808, related to #3530

The division can't be factored, the param wasn't passed through when floordiv called itself. I see this caused a number of reverts, apologies.

### CODE REVIEW

File: tinygrad/shape/symbolic.py [hunk 1]
```diff
@@ -51,12 +51,12 @@ def __rmul__(self, b:int): return self*b
   def __rfloordiv__(self, b:int): return NumNode(b) // self
   def __floordiv__(self, b:Union[Node,int], factoring_allowed=True):
     if isinstance(b, Node):
-      if b.__class__ is NumNode: return self // b.b
+      if b.__class__ is NumNode: return self.__floordiv__(b.b, factoring_allowed)
```
Comment (patosai) [inline tinygrad/shape/symbolic.py:54]: the problem was something like `Node.__floordiv__(NumNode(-16), factoring_allowed=False)` was being called, this ran and called itself but didn't give factoring_allowed so it ended up being `Node.__floordiv__(-16, factoring_allowed=True)`

Comment (chenyuxyz) [reply tinygrad/shape/symbolic.py:54]: Can you add a direct test for this case in test/unit/test_symbolic.py?

Comment (patosai) [reply tinygrad/shape/symbolic.py:54]: yup

Comment (patosai) [reply tinygrad/shape/symbolic.py:54]: test is added now, you can see the subtleness of the bug

Comment (chenyuxyz) [reply tinygrad/shape/symbolic.py:54]: thanks. yea that looks very sketchy, we probably have other bugs around div negative numbers. I will think about how to improve that.

change looks good. I will run more tests tomorrow and merge the fix if all looks good!

**Comment by chenyuxyz** [comment]
Tested with local fuzzer and ran on ci benchmark 3 times, all looked good. Thanks for fixing the bug!

**Pull request merged**